cps/elf.c

```diff
@@ -0,0 +1,261 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <elf.h>
+
+#include "gatekeeper_cps.h"
+#include "elf.h"
+
+#if __WORDSIZE == 32
+#define PERBIT(x) x##32
+#define ElfPERBIT(x) Elf32_##x
+#else
+#define PERBIT(x) x##64
+#define ElfPERBIT(x) Elf64_##x
+#endif
+
+struct elf_file {
+	/* File contents. */
+	void *data;
```

Comment:
There becomes an issue here when trying to convert the endianness of pieces of the file whose contents we've specified as `const`. Lines like this:

```
e_shnum = END(hdr->e_shnum, conv);
```

cause a problem because the `END()` macro creates a copy of the first parameter passed in using `typeof()`:

```
typeof(x) __x;
```

Compilation then fails because `__x` is marked as being `const`, so passing it into `__swap_bytes()` as the destination buffer won't work.

In other words, I think the issue is that we can't un-`const` the type passed to `typeof()`. We could create writable buffers of the same type to pass into `END()`, but it might be a little clunky. Thoughts?

Reply:
Your solution was perfect.